### PR TITLE
Port `std::span` and enable if for C++11 onwards to support mdspan

### PR DESCRIPTION
**Pull request by miscco**
In addition to porting the span implementation from libc++ the PR does the following:

* Remove outdated support for `tuple` interface
* Remove outdated support for `const_iterator`
* Change `index_type` to `size_type`

I intentionally did not adopt the ranges support, as that is out of scope.

### CODE REVIEW

**Comment by miscco** [comment]
Notably, this does not include any of the ranges work. We can only add that once we added ranges support.

File: include/cuda/std/detail/libcxx/include/type_traits [hunk 1]
```diff
@@ -4767,6 +4767,8 @@ struct __has_operator_addressof
 
 #endif  // _LIBCUDACXX_CXX03_LANG
 
+template <class...> using _Void_t = void;
```
Comment (jrhemstad) [inline include/cuda/std/detail/libcxx/include/type_traits:4770]: can we just change the `#if _LIBCUDACXX_STD_VER > 11` below as well? Those are already backported from C++17 to C++14. Might as well go to 11 if we can.

Comment (miscco) [reply include/cuda/std/detail/libcxx/include/type_traits:4770]: yeah that sounds like a good idea

Comment (miscco) [reply include/cuda/std/detail/libcxx/include/type_traits]: actually scratch that. It is only used for container like types. And there the issue is that we rely on `cuda::std::data` and `cuda::std::size` which are only c++17.

So we do not even need to change that at all unless we want to also backport those

File: include/cuda/std/detail/libcxx/include/span [hunk 2]
```diff
@@ -32,42 +39,39 @@ template <class ElementType, size_t Extent>
...
 
-namespace std {
 template <class ElementType, size_t Extent = dynamic_extent>
 class span {
 public:
     // constants and types
     using element_type = ElementType;
     using value_type = remove_cv_t<ElementType>;
-    using index_type = size_t;
+    using size_type = size_t;
```
Comment (jrhemstad) [inline include/cuda/std/detail/libcxx/include/span:48]: I'm surprised by the changes to the synopsis docs. Was `span` implementation in libc++ out of date or something?

**Review by jrhemstad — COMMENTED** [review]
It seems like there are more changes to the upstream implementation than I'm used to seeing for porting a header for libc++. Could you update the PR description with a summary of what needed to be changed?

**Comment by miscco** [comment]
> It seems like there are more changes to the upstream implementation than I'm used to seeing for porting a header for libc++. Could you update the PR description with a summary of what needed to be changed?

Yeah, `<span>` underwent a *ton* of last minute changes. If I remember the last 3 standards meetings where quite loaded with changes. That involved:
* Removing tuple support
* Removing const_iterator support
* Changing index_type to size_type
* Ranges support

I intentionally did not adopt any ranges changes, as we do not have a ranges implementation working yet.

File: include/cuda/std/detail/libcxx/include/span (hunk 2 above)
Comment (miscco) [reply include/cuda/std/detail/libcxx/include/span:48]: Yes, there were most notably

- P1872R0 span Should Have size_type, Not index_type
- P2116R0 Removing tuple-Like Protocol Support From Fixed-Extent span

I dont remember which one removed the const_iterators

**Comment by miscco** [comment]
> > Yeah, `<span>` underwent a _ton_ of last minute changes.
> 
> Ah, so this is because our upstream copy of libc++ is a bit old. That makes more sense.
> 
> Would it be possible/sane to just checkout the `span` related files and tests from upstream libc++? Or is that effectively what you did?

Yes and no. I went ahead and ported all fixes that apply to us. However, the container interface has changed drastically due to ranges support, so it is not a 100% port.

**Comment by wmaxey** [comment]
Should we consider keeping the libcxx tests in `libcxx/test/std/...`?

I think it's an important piece of coverage.

**Comment by miscco** [comment]
> Should we consider keeping a copy of the original libcxx tests in `libcxx/test/std/...`?
> 
> I think it's an important piece of coverage.

I did not test the old tests, some of them will definitely fail, like those for the tuple interface.

So I guess we should either remove them or adopt them to the new state of things

**Comment by wmaxey** [comment]
> I did not test the old tests, some of them will definitely fail, like those for the tuple interface.
> So I guess we should either remove them or adopt them to the new state of things
We could skip interfaces we don't support. 🤔

File: include/cuda/std/detail/libcxx/include/span [hunk 3]
```diff
@@ -278,277 +279,232 @@ public:
     constexpr auto subspan() const noexcept
         -> span<element_type, _Count != dynamic_extent ? _Count : _Extent - _Offset>
     {
-        static_assert(_Offset <= _Extent, "Offset out of range in span::subspan()");
-        return {data() + _Offset, _Count == dynamic_extent ? size() - _Offset : _Count};
+        static_assert(_Offset <= _Extent, "span<T, N>::subspan<Offset, Count>(): Offset out of range");
+        static_assert(_Count == dynamic_extent || _Count <= _Extent - _Offset, "span<T, N>::subspan<Offset, Count>(): Offset + Count out of range");
+
+        using _ReturnType = span<element_type, _Count != dynamic_extent ? _Count : _Extent - _Offset>;
```
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/span:285]: We are now repeating ourselves - can we either roll back this part of the change, or remove the trailing return type from the declaration?

File: include/cuda/std/detail/libcxx/include/span [hunk 4]
```diff
@@ -163,81 +163,82 @@ template <class _Tp, size_t _Sz>
 struct __is_std_array_impl<array<_Tp, _Sz>> : public true_type {};
 
 template <class _Tp>
-struct __is_std_array : public __is_std_array_impl<remove_cv_t<_Tp>> {};
+struct __is_std_array : public __is_std_array_impl<typename remove_cv<_Tp>::type> {};
 
+#if _LIBCUDACXX_STD_VER > 14
```
Comment (griwes) [inline include/cuda/std/detail/libcxx/include/span:168]: Why is this C++17+ only?

**Review by griwes — APPROVED** [review]
LGTM overall, but please address the two comments.

File: include/cuda/std/detail/libcxx/include/span [hunk 5]
```diff
@@ -118,32 +121,29 @@ template<class T, size_t N>
...
 #include <cstddef>      // for byte
+#endif // __cuda_std__
 
 #if defined(_LIBCUDACXX_USE_PRAGMA_GCC_SYSTEM_HEADER)
 #pragma GCC system_header
 #endif
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
-#if _LIBCUDACXX_STD_VER > 17
```
Comment (miscco) [inline include/cuda/std/detail/libcxx/include/span:144]: If I remember correctly, mdspan is C++14 only, so maybe we want to restrict this also to C++14?

File: include/cuda/std/detail/libcxx/include/span (hunk 4 above)
Comment (miscco) [reply include/cuda/std/detail/libcxx/include/span:168]: The container overloads that use this specialization require `std::data` and `std::size` which are both only available in C++17

We can move them up to C++14 of course but I wanted to discuss this and forgot to write the comment

File: include/cuda/std/detail/libcxx/include/span (hunk 3 above)
Comment (miscco) [reply include/cuda/std/detail/libcxx/include/span:285]: That is coming from the libc++ implementation. I would rather stay as close as possible to that implementation if possible.

Shout if you feel strongly about that change

Comment (griwes) [reply include/cuda/std/detail/libcxx/include/span:285]: ack

File: include/cuda/std/detail/libcxx/include/span (hunk 4 above)
Comment (griwes) [reply include/cuda/std/detail/libcxx/include/span:168]: Oh that's very silly. I think we should backport them - they don't rely on any core language things that are c++17, right?

Comment (miscco) [reply include/cuda/std/detail/libcxx/include/span]: Done

File: include/cuda/std/detail/libcxx/include/span (hunk 5 above)
Comment (jrhemstad) [reply include/cuda/std/detail/libcxx/include/span:144]: Yeah, I don't think we need to go through any efforts to backport this (or anything else) to C++11.

**Comment by youyu3** [comment]
Need to update `include/cuda/std/detail/libcxx/include/version` and define `__cpp_lib_span`.  Otherwise, this worked for my `mdspan` tests.  Thanks!

**Comment by wmaxey** [comment]
Will run tests and merge if no issues occur.

**Comment by wmaxey** [comment]
> Need to update `include/cuda/std/detail/libcxx/include/version` and define `__cpp_lib_span`. Otherwise, this worked for my `mdspan` tests. Thanks!

Should that be addressed in this PR?

**Comment by griwes** [comment]
It should.

**Review by jrhemstad — APPROVED** [review]

File: .upstream-tests/test/std/views/views.span/span.cons/span.fail.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+//===----------------------------------------------------------------------===//
+// UNSUPPORTED: c++03, c++11
+
```
Comment (wmaxey) [inline .upstream-tests/test/std/views/views.span/span.cons/span.fail.cpp:10]: .fail tests do not yet support NVRTC.

File: .upstream-tests/test/std/views/views.span/span.cons/iterator_len.verify.cpp [hunk 7]
```diff
@@ -0,0 +1,38 @@
+//===---------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+//===---------------------------------------------------------------------===//
+// UNSUPPORTED: c++03, c++11
+
```
Comment (wmaxey) [inline .upstream-tests/test/std/views/views.span/span.cons/iterator_len.verify.cpp:10]: .verify tests probably also do not support NVRTC.

Comment (wmaxey) [reply .upstream-tests/test/std/views/views.span/span.cons/iterator_len.verify.cpp:10]: But I don't see any complaints in CI.

**Review by wmaxey — CHANGES_REQUESTED** [review]

**Review by wmaxey — APPROVED** [review]

**Pull request merged**